src/graph.rs

```diff
@@ -93,22 +96,53 @@ impl<'tree> Graph<'tree> {
 
     pub fn display_json<'a>(&'a self, ctx: &'a Context) -> () {
         struct JSONGraph<'a, 'tree>(&'a Graph<'tree>, &'a Context);
+        struct JSONNode<'a>(usize, &'a GraphNode, &'a Context);
+        struct JSONEdge<'a>(&'a u32, &'a Edge, &'a Context);
+
         impl<'a, 'tree> ser::Serialize for JSONGraph<'a, 'tree> {
             fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
                 where S: serde::Serializer {
```

Comment:
Out of curiosity, is `where S` simply a way to get local abbreviation here, or does this do anything else?

Reply:
In Rust you express trait constraints not directly within the signature, but in a `where` clause afterwards. I’m not sure if this is a hard language-level requirement or just a general practice, but you see it all over the stdlib.